cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,305 @@
+"""
+``remove_fixable_singularities`` specifies a method to remove fixable singularities from a model
+"""
+from math import isclose
+
+import networkx as nx
+from sympy import (
+    Abs,
+    Add,
+    Eq,
+    Float,
+    Mul,
+    Piecewise,
+    Pow,
+    S,
+    Wild,
+    exp,
+    log,
+    solveset,
+)
+from sympy.codegen.rewriting import ReplaceOptim, optimize
+from sympy.sets import Intersection
+
+from .model import FLOAT_PRECISION, Quantity
+
+
+# For P^n make sure n is passed as int if it is actually a whole number
+_POW_OPT = ReplaceOptim(lambda p: p.is_Pow and (isinstance(p.exp, Float) or isinstance(p.exp, float))
+                        and float(p.exp).is_integer(),
+                        lambda p: Pow(p.base, int(float(p.exp))))
+
+
+ONE = Quantity(1.0, 'dimensionless')
+
+
+def _generate_piecewise(expr, V, sp, Vmin, Vmax):
+    """
+    Generates a new (piecewise) expression based on expr with a linear interpolation when V is between Vmin and Vmax
+
+    The returned expression between Vmin and Vmax is::
+    f(Vmin) + (V - Vmin) / (Vmax - Vmin) * (f(Vmax) - f(Vmin))
+where ``f`` is ``expr``
+
+    :param expr: The expression to turn into a piecewise
+    :param V: The voltage variable.
+    :param sp: The value of the singularity point: the value of V for which U == 0.
+    :param Vmin: The value of the lower bound: the value of V for which U - U_offset == 0.
+    :param Vmax: The value of the upper bound: the value of V for which U + U_offset == 0.
+    :return: expr with a singularity exception around sp, if sp is not None else return the original expr.
+    """
+    f_Vmin = expr.xreplace({V: Vmin})
+    f_Vmax = expr.xreplace({V: Vmax})
+    return Piecewise((f_Vmin + ((V - Vmin) / (Vmax - Vmin)) * (f_Vmax - f_Vmin),
+                      Abs(V - sp) < Abs((Vmax - Vmin) / 2)), (expr, True))
+
+
+def _float_dummies(expr):
+    """Turns floats back into Quantity dummies to be in line with the rest of the sympy equations"""
+    return expr.xreplace({f: Quantity(f, 'dimensionless') for f in expr.atoms(Float)})
+
+
+def _is_negative_power(expr):
+    try:
+        return isinstance(expr, Pow) and bool(expr.args[1] < 0)
+    except TypeError:  # if this is a power with variables still in it we can't determine if it's negative
+        pass
+    return False
+
+
+def _solve_real(u, V):
+    """Gives the values of V for which u == 0 """
+    u = optimize(u, (_POW_OPT, ))  # make sure powers of ints are represented as ints
+    result = solveset(u, V, domain=S.Reals)
+    # The resul could be an intersection with Reals, if a custom exp function is used. We assume the result is real.
+    if isinstance(result, Intersection) and result.args[0] == S.Reals:
+        result = result.args[1]
+    return result
+
+
+def _get_singularity(expr, V, U_offset, exp_function):
+    """
+    Finds singularities in equations of the form:
+       - ``U / (exp(U) - 1.0)``
```

Comment:
Further down we look for `exp(U) * -Z + 1.0`. Sounds like that should be mentioned here?

Reply:
actually `exp(U) * -Z + 1.0` 
is
`Z * ((1.0 - exp(U)) / U)`
so that's alradymentioned?